src/qibo/noise.py

```diff
@@ -173,12 +175,31 @@ def add(self, error, gate, qubits=None):
             gate (:class:`qibo.gates.Gate`): gate after which the noise will be added.
             qubits (tuple): qubits where the noise will be applied, if None the noise
                             will be added after every instance of the gate.
+            condition (callable): Optional function that takes :class:`qibo.gates.Gate` object as an input and returns True if noise should be added to it.
         """
 
         if isinstance(qubits, int):
             qubits = (qubits,)
 
-        self.errors[gate] = (error, qubits)
+        if condition is not None:
+            if callable(condition):
+                if gate not in self.conditions:
+                    self.conditions[gate] = ([condition], [error], [qubits])
+                else:
+                    self.conditions[gate][0].append(condition)
+                    self.conditions[gate][1].append(error)
+                    self.conditions[gate][2].append(qubits)
+            else:
+                raise TypeError(
+                    "condition should be callable. Got {} instead."
+                    "".format(type(condition))
+                )
+        else:
+            if gate not in self.errors:
+                self.errors[gate] = ([error], [qubits])
+            else:
+                self.errors[gate][0].append(error)
+                self.errors[gate][1].append(qubits)
```

Comment:
Here something like:
```py
self.errors[gate].append((error, qubits))
```
Btw, in this approach if you initialize as `self.errors = collections.defaultdict(list)` instead of the standard dict you also don't need the `if gate not in self.errors` check.